sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,821 @@
+        self.X_n_samples_remainder = self.n_samples_X % self.X_n_samples_chunk
+
+        self.n_samples_Y = datasets_pair.n_samples_Y()
+        self.Y_n_samples_chunk = min(self.n_samples_Y, self.chunk_size)
+        Y_n_full_chunks = self.n_samples_Y // self.Y_n_samples_chunk
+        self.Y_n_samples_remainder = self.n_samples_Y % self.Y_n_samples_chunk
+
+        # Counting remainder chunk in total number of chunks
+        self.X_n_chunks = X_n_full_chunks + (self.X_n_samples_remainder != 0)
+        self.Y_n_chunks = Y_n_full_chunks + (self.Y_n_samples_remainder != 0)
+
+        if strategy is None:
+            strategy = get_config().get("pairwise_dist_parallel_strategy", 'auto')
+
+        if strategy not in ('parallel_on_X', 'parallel_on_Y', 'auto'):
+            raise RuntimeError(f"strategy must be 'parallel_on_X, 'parallel_on_Y', "
+                               f"or 'auto', but currently strategy='{self.strategy}'.")
+
+        if strategy == 'auto':
+            # This is a simple heuristic whose constant for the
+            # comparison has been chosen based on experiments.
+            if 4 * self.chunk_size * self.effective_n_threads < self.n_samples_X:
+                strategy = 'parallel_on_X'
+            else:
+                strategy = 'parallel_on_Y'
+
+        self.execute_in_parallel_on_Y = strategy == "parallel_on_Y"
+
+        # Not using less, not using more.
+        self.chunks_n_threads = min(
+            self.Y_n_chunks if self.execute_in_parallel_on_Y else self.X_n_chunks,
+            self.effective_n_threads,
+        )
+
+    @final
+    cdef void _parallel_on_X(self) nogil:
+        """Compute the pairwise distances of each vector (row) of X on Y
+        by parallelizing computation on chunks of X and reduce them.
+
+        This strategy dispatches chunks of X uniformly on threads.
+        Each thread then iterates on all the chunks of Y. This strategy is
+        embarrassingly parallel and comes with no datastructures synchronisation.
+
+        Private datastructures are modified internally by threads.
+
+        Private template methods can be implemented on subclasses to
+        interact with those datastructures at various stages.
+        """
+        cdef:
+            ITYPE_t Y_start, Y_end, X_start, X_end, X_chunk_idx, Y_chunk_idx
+            ITYPE_t thread_num
+
+        with nogil, parallel(num_threads=self.chunks_n_threads):
+            thread_num = _openmp_thread_num()
+
+            # Allocating thread datastructures
+            self._parallel_on_X_parallel_init(thread_num)
+
+            for X_chunk_idx in prange(self.X_n_chunks, schedule='static'):
+                X_start = X_chunk_idx * self.X_n_samples_chunk
+                if (X_chunk_idx == self.X_n_chunks - 1
+                    and self.X_n_samples_remainder > 0):
+                    X_end = X_start + self.X_n_samples_remainder
+                else:
+                    X_end = X_start + self.X_n_samples_chunk
```

Comment:
This is a pattern that occurs 4 times in the code. Can you make an inline cdef helper function for it?

Reply:
I think it's easier keeping this as is: the logic is simple enough and introducing a helper method for this would be uneasy as two values would need to be returned.

Reply:
In order to return a pair of values, have a look at https://github.com/scikit-learn/scikit-learn/blob/5bb18166d420c5e265844c32442d960ca0c5005e/sklearn/_loss/_loss.pxd#L21-L24 and then a function like `cy_grad_hess`.

I thought, it would save a little code and make reading it easier. No hard feelings, if you won't pursue this suggestion.

Reply:
Let's wait for someone else's opinion.

Reply:
I think we can do a slight refactor of `__init__` to not need a helper here:

```python
X_n_samples_remainder = self.n_samples_X % self.X_n_samples_chunk
self.X_n_chunks = X_n_full_chunks + (X_n_samples_remainder != 0)

if X_n_samples_remainder != 0:
    self.X_n_samples_last_chunk = X_n_samples_remainder
else:
    self.X_n_samples_last_chunk = X_n_samples_chunk
```

In the for loop we can now have:

```python
for X_chunk_idx in prange(self.X_n_chunks, schedule='static'):
    X_start = X_chunk_idx * self.X_n_samples_chunk
    if X_chunk_idx == self.X_n_chunks - 1:
        X_end = X_start + self.X_n_samples_last_chunk
    else:
        X_end = X_start + self.X_n_samples_chunk
```

I think only checking for the last index is simple enough to keep inline.

Reply:
Addressed in a2f7b6d11597388e95443325d692037ef6eb321c.

Please, let me know what you think of it.

Reply:
> Please, let me know what you think of it.

LGTM!